### PR TITLE
Optimizing expressions on empty PREFIX/CONTAINS/SUFFIX needles

### LINKED ISSUE #572: Query using PREFIX() results in an error std::bad_alloc
Consider the following statements:
```sql
CREATE TABLE t0(c0 VARCHAR);
SELECT * FROM t0 WHERE PREFIX(t0.c0, ''); -- Error: std::bad_alloc
```
Unexpectedly, the `SELECT` results in an error `Error: std::bad_alloc`. I found this based on commit a09d2f484db83ac9e27c9501f439bcd32b0083ec.

### PR DESCRIPTION
**Pull request by hannes**
Added rewrite rule that optimises
`PREFIX(x, '')`, `CONTAINS(x, '')` and `SUFFIX(x, '')` to `CASE WHEN x IS NOT NULL THEN TRUE ELSE NULL END`
also fixes issue #572

### CODE REVIEW

File: src/optimizer/filter_combiner.cpp [hunk 1]
```diff
@@ -239,6 +239,9 @@ FilterCombiner::GenerateTableScanFilters(std::function<void(unique_ptr<Expressio
 				auto &column_ref = (BoundColumnRefExpression &)*func.children[0].get();
 				auto &constant_value_expr = (BoundConstantExpression &)*func.children[1].get();
 				string like_string = constant_value_expr.value.str_value;
+				if (like_string.size() == 0) {
```
Comment (pdet) [inline src/optimizer/filter_combiner.cpp:242]: Makes sense :-), I think .empty() is a bit cleaner though

Comment (pdet) [reply src/optimizer/filter_combiner.cpp:242]: However, shouldn't this case be optimized away?

Comment (hannesmuehleisen) [reply src/optimizer/filter_combiner.cpp:242]: > However, shouldn't this case be optimized away?

Indeed it should. But the check still needs to be there.

**Comment by hannesmuehleisen** [comment]
@pdet I have added an optimiser as suggested.

File: src/optimizer/rule/empty_prefix_removal.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+using namespace duckdb;
+using namespace std;
+
+EmptyPrefixRemovalRule::EmptyPrefixRemovalRule(ExpressionRewriter &rewriter) : Rule(rewriter) {
+	// match on a FunctionExpression that has a foldable ConstantExpression
+	auto func = make_unique<FunctionExpressionMatcher>();
+	func->matchers.push_back(make_unique<ExpressionMatcher>());
+	func->matchers.push_back(make_unique<ExpressionMatcher>());
+	func->policy = SetMatcher::Policy::SOME;
+	func->function = make_unique<SpecificFunctionMatcher>("prefix");
```
Comment (Mytherin) [inline src/optimizer/rule/empty_prefix_removal.cpp:18]: Perhaps also handle SUFFIX/CONTAINS here?

Comment (hannesmuehleisen) [reply src/optimizer/rule/empty_prefix_removal.cpp:18]: that's a great idea. I had no idea these functions existed :)

**Pull request merged**